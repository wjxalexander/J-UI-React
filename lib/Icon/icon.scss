.J-UI-Icon{
  width: 1em;
  height: 1em;
  font-size: 4em;

}