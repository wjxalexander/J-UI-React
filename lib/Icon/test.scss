.J-UI-Icon{
  color:red;
  width: 1em;
  height: 1em;
  fill: red;
}