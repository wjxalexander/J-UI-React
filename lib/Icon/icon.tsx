import * as React from "react";
import {Fragment} from "react";
import cls from "classnames";
import "./utils/importAllIcons.js";
import "./fabicIcons/fabric-icons-inline.scss"
import "./icon.scss"

/*所有的Onxx 响应全部在DOMAttributes<T>中 extends里面了  可点击看react源码 这里因为会是SVG或者html元素 因此元素<T>是HTMLOrSVGElement
 * 最终传给DOMAttributes<T>
 */
interface IconProps extends React.SVGAttributes<HTMLOrSVGElement>, React.HTMLAttributes<HTMLOrSVGElement> {
    iconName: string;
    className?: string;
    USEMsFabricIcon?: boolean
}

//  用户可以使用自己的iconfont
const customCache = new Set<string>();

export function createFromIconFont(scripts: string = ""): void {
    if (
        typeof document !== 'undefined' &&
        typeof window !== 'undefined' &&
        typeof document.createElement === 'function' &&
        scripts.length &&
        !customCache.has(scripts)
    ) {
        const script = document.createElement('script');
        script.setAttribute('type', "text/javascript");
        script.setAttribute('src', scripts);
        script.setAttribute('script-import-from-user', scripts);
        customCache.add(scripts);
        document.body.appendChild(script);
    }
}

// 为什么在JSX中使用...运算符需要加{} 因为他是JS语法，JSX规定需要加{}
const SvgComponent: React.FunctionComponent<IconProps> = (({iconName, className, ...restProps}) => {
    return (
        <svg className={cls("J-UI-Icon", className)} {...restProps}>
            <use xlinkHref={`#${iconName}`}/>
        </svg>)
});

const HTMLComponent: React.FunctionComponent<IconProps> = (({iconName, className, ...restProps}) => {
    return <i className={cls(`ms-Icon ms-Icon--${iconName} J-UI-Icon`, className)}{...restProps}/>
});

// Icon 是一个React的函数式组件，他的类型是IconProps, props:IconProps+children
const Icon: React.FunctionComponent<IconProps> = ({USEMsFabricIcon, ...restProps}) => {
    const DisplayComponent = USEMsFabricIcon ? (<HTMLComponent {...restProps}/>) : (<SvgComponent {...restProps}/>);
    return (
        <Fragment>
            {DisplayComponent}
        </Fragment>
    )
};
export default Icon;