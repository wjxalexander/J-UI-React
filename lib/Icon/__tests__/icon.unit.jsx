import renderer from 'react-test-renderer'
import Icon,{createFromIconFont} from "../icon"
import React from 'react'; // !!!
import {mount} from "enzyme"

describe("icon test", () => {
    it("is a svg", () => {
        const tree = renderer
            .create(<Icon iconName="wechat"/>)
            .toJSON() //虚拟dom是个对象
        expect(tree).toMatchSnapshot()
    })
    it("is a HTML", () => {
        const tree = renderer
            .create(< Icon iconName="AdminALogo32" USEMsFabricIcon/>)
            .toJSON() //虚拟dom是个对象
        expect(tree).toMatchSnapshot()
    });
    // renderer只用于snapshot
    it("onClick svg", () => {
        const fakeFn = jest.fn();
        const component = mount(<Icon iconName="wechat" onClick={fakeFn}/>)
        component.find('svg').simulate('click')
        expect(fakeFn).toBeCalled()
    });
    it("onClick fabric", () => {
        const fakeFn = jest.fn()
        const component = mount(<Icon iconName="AdminALogo32" USEMsFabricIcon onClick={fakeFn}/>)
        component.find('i').simulate('click')
        expect(fakeFn).toBeCalled()
    });
    it("onMouseLeave svg", () => {
        const fakeFn = jest.fn()
        const component = mount(<Icon iconName="wechat" onMouseLeave={fakeFn}/>)
        component.find('svg').simulate('mouseleave')
        expect(fakeFn).toBeCalled()
    });
    it("onMouseLeave svg", () => {
        const fakeFn = jest.fn()
        const component = mount(<Icon iconName="wechat" onMouseLeave={fakeFn}/>)
        component.find('svg').simulate('mouseleave')
        expect(fakeFn).toBeCalled()
    });
    it("onTouchCancel svg", () => {
        const fakeFn = jest.fn()
        const component = mount(<Icon iconName="wechat" onTouchCancel={fakeFn}/>);
        component.find('svg').simulate('touchcancel');
        expect(fakeFn).toBeCalled()
    });
    it("outside svg", () => {
        createFromIconFont("http://at.alicdn.com/t/font_891885_dszg7ff4by8.js")
        const tree = renderer
            .create(<Icon iconName="icon-history"/>)
            .toJSON() //虚拟dom是个对象
        expect(tree).toMatchSnapshot()
    });
    it("outside svg onTouchCancel", () => {
        createFromIconFont("http://at.alicdn.com/t/font_891885_dszg7ff4by8.js")
        const fakeFn = jest.fn()
        const component = mount(<Icon iconName="icon-history" onTouchCancel={fakeFn}/>);
        component.find('svg').simulate('touchcancel');
        expect(fakeFn).toBeCalled()
    });
});
