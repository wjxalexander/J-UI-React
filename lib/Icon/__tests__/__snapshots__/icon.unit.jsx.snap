// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`icon test MSFabric is a HTML 1`] = `
<i
  className="ms-Icon ms-Icon--AdminALogo32 J-UI-Icon"
/>
`;

exports[`icon test is a svg 1`] = `
<svg
  className="J-UI-Icon"
>
  <use
    xlinkHref="#wechat"
  />
</svg>
`;
